expo/expo
# [babel-preset-expo] Added support for jsxRuntime option

# Why

Now that we support React 17, we can add support for their big feature: [auto JSX transforms](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html).



# How


The React Native preset does not expose this functionality, so I opted to disable the transforms entirely, and add them back with customizations.

I documented my findings in-depth since Babel can be a bit confusing.

## Misc Guides

### Eslint

To disable eslint warnings, create a `.eslintrc.js` and add:

```js
module.exports = {
  rules: {
    // Disable React JSX transform warnings because
    // `jsxRuntime: 'automatic'` is passed to `babel-preset-expo`.
    'react/jsx-uses-react': 'off',
    'react/react-in-jsx-scope': 'off',
  },
};
```

### TypeScript

> Requires TypeScript 4.1

In your `tsconfig.json`, add the [new jsx setting](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#jsx-factories):

```json
{
  "compilerOptions": {
    "jsx": "react-jsx"
  },
  "extends": "expo/tsconfig.base"
}
```

### Flow

[idk](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#flow).



# Test Plan

- Wrote unit tests, but pending https://github.com/expo/expo/pull/13946

- Run `expo start -c` in bare-sandbox, it should load up as expected.
- Disable the babel `jsxRuntime` option, and try again, it should produce an error "Can't find variable: React", this is because I removed React from the App.js component to test it.

# Checklist
